service/storage_proxy.cc

```diff
@@ -1386,28 +1386,11 @@ class abstract_write_response_handler : public seastar::enable_shared_from_this<
                     return std::max(lhs, rhs);
                 });
     }
-    std::chrono::microseconds calculate_delay(db::view::update_backlog backlog) {
-        constexpr auto delay_limit_us = 1000000;
-        auto adjust = [] (float x) { return x * x * x; };
-        std::chrono::microseconds ret(uint32_t(adjust(backlog.relative_size()) * delay_limit_us));
-
-        // The maximum delay applied to the response is the amount of time left until the request timeouts.
-        // Delaying the response more than the timeout point doesn't make much sense - the request will fail
-        // at the timeout point anyway.
-        std::chrono::nanoseconds budget =
-            std::max(std::chrono::nanoseconds(0), _expire_timer.get_timeout() - storage_proxy::clock_type::now());
-
-        // budget is in nanoseconds, so it can be cast to microseconds without worrying about overflows,
-        // casting from nano to micro means dividing by 1000.
-        std::chrono::microseconds max_delay = std::chrono::duration_cast<std::chrono::microseconds>(budget);
-
-        return std::min(ret, max_delay);
-    }
```

Comment:
I don't know if we want to reuse the sleep-based throttling here.

For the client, we can't control its concurrency, so we just slow it down.

But for view update generation, isn't it enough to limit it with a semaphore?

Reply:
@avikivity, how would you use a semaphore?

Imagine that there is just **one** delete going on, that wants to delete a million rows. After deleting a batch 100 rows, how quickly should the deletion send another batch of 100 rows?  If it sends the next batch immediately, it overloads the server with background work (that's the current situation).

I guess it could somehow wait until these 100 specific view rows were written, but why 100? The idea of the sleep is to use the same flow control algorithm that an ordinary "deletion client" with concurrency 100 would have used. This deletion client may send much more than 100 concurrent writes as a burst, but as soon as the backlog begins to grow, it will start slowing down.

Reply:
We could limit the amount of concurrent `view updates from range delete` operations, but this causes the usual issues with limiting concurrency - what number of concurrent requests is the right one - 100, 10000, 1e6?. With high latency we want the window size to be large, with low latency it should be small.

This solution treats each range delete as a INSERT workload with concurrency of 100, and uses the current backpressure mechanism to limit the number of concurrent updates.

IMO ideally it should look like this:
* The node that generates the updates pushes them onto a TCP connection, blocking until the `send()` completes.
* The receiver node reads as many updates as it can reasonably handle at once (say, 1000 concurrent requests)
* TCP flow control takes care of choosing the right amount of in-flight requests

But I think this would require some changes to RPC. I have some thoughts about changing how the backpressure works, I'll try to make a clear document about it.

For now I feel that this is the best solution - it uses the existing mechanism and fixes the issue.

Reply:
We need to maintain some concurrency limit. In the worse case, every single shard is performing view updates, so the total concurrency is one_shard_concurrency * nr_shards. In the worst case all of this concurrency is aimed at one shard (which would cause immense slowdowns, but that's life). In the best case, it's spread evenly across the cluster.

I think shard-wide (or service-level-wide) concurrency of 100 is reasonable. And above that we delay the client (by delaying the coordinator).

With sleeping, the concurrency just keeps growing. Yes, we're sending fewer requests, but our own concurrency is unbounded.

Reply:
> > With sleeping, the concurrency just keeps growing. Yes, we're sending fewer requests, but our own concurrency is unbounded.
> 
> Why does our concurrency just keep growing? We're (the base replica processing the partition delete) are sending fewer requests, and not doing anything else. We're sending batches of row deletes 100 at a time, and don't have any background work besides the work we send to other nodes, so how can **our** concurrency be growing? The concurrency on those other nodes can indeed grow with all that background work, but that concurrency we check - and if it grows - we increase the sleep.

We're processing new incoming replica writes at the same time.

Let's say we get a delete from a 1 GB partition. We start the send/sleep/send thing for that partition.
A millisecond later, another 1GB delete.
After a second, we're asked to delete 1000 such partitions. We have a concurrency of 1000.

Reply:
The whole idea of batching is wrong. We should have a concurrency limit, but why batch? read a row, process it (background), read another row. process it (background), eventually you hit the concurrency limit and wait until some row completes.

Would have solved the issues with range tombstones that span batch boundaries too.

Reply:
100ms is not typical latency for a base replica and a view replica.

Reply:
And we can increase 100 to 1000, just as long as it's bounded. Changing from one unbounded algorithm to another isn't good.

Reply:
The view updates are supposed to be asynchronous - we spawn as many of them as will fit in the queue, and then process them in the background. The only waiting we should be doing is waiting for a place in the queue.
Bounding this to 100 requests goes against this idea, it will make all such view updates synchronous.

Reply:
> > Let's say we get a delete from a 1 GB partition. We start the send/sleep/send thing for that partition. A millisecond later, another 1GB delete. After a second, we're asked to delete 1000 such partitions. We have a concurrency of 1000.
> 
> If the user sends 1000 deletions of 1GB partitions concurrently, there will indeed be a lot of concurrent work - the internal concurrency will be 1000*100. But is this unexpected?

Why not?

> The user knows these are huge deletions, they should expect that doing 1000 of them in parallel should be difficult.

Why would they know that? And why would they care? The database shouldn't crash because you throw something difficult at it.

> Not impossible - concurrency of 100,000 is not out of the question. Anyway, what is the alternative?

Throttle down the incoming base replica work.

> If you have 1000 concurrent partition deletions, and each needs to process 1 million row deletions, would you expect each partition deletion to process and send one row at a time? Today we have the batch size 100 just to make the reading part more efficient, would we want to lower the batch size if there are a lot of concurrent partition deletions? I'm not sure.

No. Limit each service level's view update concurrency to something reasonable (100-1000). Once it's exceeded, require progress to be made on the view updates before processing more base updates.

The batching is wrong, we already have readers that do batching. The extra batching on top is a source of bugs and inefficiency.

Reply:
> The view updates are supposed to be asynchronous - we spawn as many of them as will fit in the queue, and then process them in the background. The "concurrency" is supposed to be sky high. The only waiting we should be doing is waiting for a place in the queue. Bounding this to 100 requests goes against this idea, it will make all such view updates synchronous.


What limits the queue size?

To have end-to-end throttling, we need to stop processing base updates (which will stop processing client requests as the back-pressure propagates).

I don't understand how the sleeps help. They spread the requests over time, but more keep coming in. If our base replica processing throughput is higher than the view update throughput, we'll run out of memory.

Reply:
@avikivity you're absolutely right that "the extra batching on top is a source of bugs" - see for example 92d03be37b407dfb48d3cf96ef28bfbe42aed3b2. But this is what happens when we keep piling on fixes on fixes on fixes on fixes without having the time and peace of mind to actually give some TLC to our code. So we already have this 100-row batching in place.

If you're curious where this 100-row-batching came from, it started by reading the whole partition sequentially, and then in bf0777e97a0abb3bebddf84c82de774ebfd9255d this 100-row batches happened.